onnxscript/rewriter/ort_fusions/attention.py

```diff
@@ -0,0 +1,283 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+from __future__ import annotations
+
+from typing import Sequence, Union
+
+import onnxscript.ir as ir
+from onnxscript.rewriter import pattern
+
+Dim = Union[int, ir.SymbolicDim]
+
+
+# TODO: Maybe add this check to utilities
+def _check_shape(bindings: dict[str, Dim], val: ir.Value, shape: Sequence[str]) -> bool:
+    if val.shape is None:
+        return False
+    if val.shape.rank() != len(shape):
+        return False
+    for actual, expected in zip(val.shape, shape):
+        if expected not in bindings:
+            bindings[expected] = actual  # type: ignore[assignment]
+        elif actual != bindings[expected]:
+            return False
+    return True
+
+
+class AttentionFusion(pattern.RewriteRuleClassBase):
+    def __init__(self, name, *, has_input_bias: bool, has_past: bool = False):
+        super().__init__(name)
+        # TODO: We can just pass bias to MultiHeadAttention
+        # and let it handle the bias addition, once that pattern is added to MHA
+        self._has_input_bias = has_input_bias
+        self._has_past = has_past
+
+    def pattern(
+        self,
+        op,
+        input,
+        qkv_weight,
+        qkv_bias,
+        # mask_index,
+        past,
+        # attention_bias,
+        num_heads,
+        # scale,
+    ):
+        projected = op.MatMul(input, qkv_weight)
+        # Add bias if present
+        if self._has_input_bias:
+            projected = op.Add(projected, qkv_bias)
+
+        # Slice packed Matmul QKV into Q, K, and V
+        # Q, K, and V are of shape (B, S, D)
+        query_BSD = op.Slice(
+            projected,
+            _allow_other_inputs=True,
+            _allow_other_attributes=True,
```

Comment:
We can drop the _allow_other_attributes, Slice doesn't have any attributes. Further, True is the default value for this flag anyway. (I realize this probably comes from a copy of pre-existing code, which was meant to be cleaned up, but wasn't).